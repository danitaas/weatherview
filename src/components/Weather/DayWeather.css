

.DayWeather-container {
  display: flex; /* todo: flex grid if target browser supports it */
  flex-direction: row;
  width: auto;
  flex-wrap: wrap;
  padding: 10px;
}

.DayWeather-item {
  width: 180px; /* todo: flex grid if target browser supports it */
  margin-bottom: 10px;
}