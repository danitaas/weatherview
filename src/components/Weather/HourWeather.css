

.HourWeather-container {
    /*width: 200px;*/
    padding: 8px;
}

.HourWeather-item {

}