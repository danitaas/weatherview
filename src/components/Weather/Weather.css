
.Weather-container {
  margin-top: 50px;
  display: flex;
  flex-direction: row;
}

.Weather-dayitems {
    flex-grow: 1;
  display: flex;
  flex-direction: column;
}

.Weather-dayitem {
  text-align: left;
  margin-bottom: 20px;
}

.Weather-day {
    min-width: 200px;
    background-color: #e2e2e2;
}

.Weather-houritems {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
}

.Weather-houritem {
    text-align: left;
    margin-bottom: 10px;
}