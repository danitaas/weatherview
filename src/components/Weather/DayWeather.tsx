import * as React from 'react';
import './DayWeather.css';

// todo: naming conventions/data formats
export interface IProps {
    day?: string;
    mintemp?: string;
    maxtemp?: string;
    conditions?: string;
    wind?: string;
}

function DayWeather(props: IProps) {
    if (props.day == "error") {
        throw new Error('An error occured');
    }
    return (
        <div className="DayWeather-container">
            <div className="DayWeather-item">
                <div>{props.day}</div>
                <div>{props.mintemp}</div>
                <div>{props.maxtemp}</div>
                <div>{props.conditions}</div>
                <div>{props.wind}</div>
            </div>
        </div>
    );
}

export default DayWeather;
