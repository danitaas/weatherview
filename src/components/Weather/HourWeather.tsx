import * as React from "react";
import "./DayWeather.css";
import { IDayWeather } from "../../logic/weather";
import * as moment from "moment";
import {
    formatCentigradeFromKelvin, formatDegToCard, formatFahrenheitFromKelvin,
    formatKMHFromMS, formatMPHFromMS
} from "../../logic/maths";

export interface IProps extends Partial<IDayWeather> {

}

function HourWeather(props: IProps) {
    if (!props.dt) {
        return <div>Invalid data</div>;
    }

    //could pre-format these in redux
    const mintempstr = props.mintemp ? formatCentigradeFromKelvin(props.mintemp) + "째C" + "/" + formatFahrenheitFromKelvin(props.mintemp) + "째F" : "Unknown";
    const maxtempstr =  props.maxtemp ? formatCentigradeFromKelvin(props.maxtemp) + "째C" + "/" + formatFahrenheitFromKelvin(props.maxtemp) + "째F" : "Unknown";
    const windspeedstr =  props.windspeed ? formatKMHFromMS(props.windspeed) + "kmh" + "/" + formatMPHFromMS(props.windspeed) + "mph" : "Unknown";
    const winddegstr =  props.windspeed ? formatDegToCard(props.windspeed) : "Unknown";
    const windstr =  windspeedstr + " " + winddegstr;

    //{JSON.stringify(i)}

    return (
        <div className="HourWeather-container">
            <div className="HourWeather-item">{moment.unix(props.dt).format("HH:mm")}</div>
            <div className="HourWeather-item">{mintempstr}</div>
            <div className="HourWeather-item">{maxtempstr}</div>
            <div className="HourWeather-item">{props.conditions}</div>
            <div className="HourWeather-item">{windstr}</div>
        </div>
    );
}

export default HourWeather;
