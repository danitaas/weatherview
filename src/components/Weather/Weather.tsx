import * as React from "react";
import { connect } from "react-redux";
import { IDayWeather, IWeatherState, SelectDay, selectDayForecasts, selectWeekForecasts } from "../../logic/weather";
import DayWeather from "./DayWeather";
import "./Weather.css";
import HourWeather from "./HourWeather";
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Paper from "@material-ui/core/Paper";
import DialogTitle from '@material-ui/core/DialogTitle';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';

export interface IProps {
    weekforecasts?: IDayWeather[];
    dayforecasts?: IDayWeather[];
    currentday?: string;
    dispatch?: any;
}

export interface IState {
    dialogopen: boolean;
}

export class Weather extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.handleClick = this.handleClick.bind(this);

        this.state = {
            dialogopen: false,
        };
    }

    private handleClick(dt: number) {
        console.log("handleClick", dt);

        this.setState({
            dialogopen: true,
        });

        //could be actioncreator injected by connect or inline
        this.props.dispatch(SelectDay(dt));
    }

    //note: ES6 auto this binding
    private handleClose = () => {
        this.setState({
            dialogopen: false,
        });
    };

    public render() {
        return (
            <div className="Weather-container">
                <div className="Weather-dayitems">
                    {this.props.weekforecasts && this.props.weekforecasts.length > 0 ? (
                        this.props.weekforecasts.map(i => {
                            return (
                                // todo: key should be record-id
                                <div key={"dayitem" + i.dt} className="Weather-dayitem">
                                    <Paper>
                                        <DayWeather {...i} />
                                        <div style={{textAlign: 'right', padding: 8}}>
                                            {/*
                                        NOTE: as per latest guidance lambdas are okay in JSX
                                        https://reactjs.org/docs/faq-functions.html#is-it-ok-to-use-arrow-functions-in-render-methods
                                        however as number of components increases, eg large lists
                                        an alternative can be to add sub-component, with bound methods, or use curried cached methods
                                        */}
                                            <Button variant={"raised"} color="secondary" onClick={() => this.handleClick(i.dt)}>Show hourly</Button>
                                        </div>
                                    </Paper>
                                </div>
                            );
                        })
                    ) : (

                        <div>
                            <Typography variant="subheading">
                                No results found
                            </Typography>
                        </div>
                    )}
                </div>
                <Dialog
                    fullWidth={false}
                    onClose={this.handleClose}
                    open={this.state.dialogopen}>
                    <DialogTitle>Daily Forecast for: {this.props.currentday}</DialogTitle>
                    <DialogContent>
                        <div className="Weather-day">
                            <div className="Weather-houritems">
                                {this.props.dayforecasts && this.props.dayforecasts.length > 0
                                    ? this.props.dayforecasts.map(i => {
                                        return (
                                            // todo: key should be record-id
                                            <div key={"houritem" + i.dt} className="Weather-houritem">
                                                <Paper>
                                                    <HourWeather {...i} />
                                                </Paper>
                                            </div>
                                        );
                                    })
                                    : null}
                            </div>
                        </div>
                    </DialogContent>
                    <DialogActions>
                        <Button variant={"raised"} onClick={this.handleClose} color="secondary">
                            Close
                        </Button>
                    </DialogActions>
                </Dialog>
            </div>
        );
    }
}

export function mapStateToProps(state: IWeatherState): IProps {
    return {
        weekforecasts: selectWeekForecasts(state),
        dayforecasts: selectDayForecasts(state),
        currentday: state.currentday,
    };
}

// separate variables for hot reloading + test
const component = Weather;
const connectedcomponent = connect<IProps, IProps, IProps>(mapStateToProps)(component);
// todo: add other HOCs here
export default connectedcomponent;
