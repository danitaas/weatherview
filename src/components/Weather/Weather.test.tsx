import * as React from "react";
import { Weather } from "./Weather";
// todo: cant use typescript import without correcting toolchain
const ReactShallowRenderer = require("react-test-renderer/shallow");
const ReactTestRenderer = require("react-test-renderer");

it("renders shallow without crashing", () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<Weather />);
    const result = renderer.getRenderOutput();
    expect(result.type).toBe("div");
});

it("renders snapshot", () => {
    const tree = ReactTestRenderer.create(<Weather />).toJSON();
    expect(tree).toMatchSnapshot();
});

it("renders snapshot with props", () => {
    const tree = ReactTestRenderer.create(<Weather weekforecasts={[]} dayforecasts={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
});

//todo: move tests with forecasts data
