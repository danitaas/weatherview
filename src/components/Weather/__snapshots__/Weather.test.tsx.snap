// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
<div
  className="Weather-container"
>
  <div
    className="Weather-dayitems"
  >
    <div>
      No results found
    </div>
  </div>
  <div
    className="Weather-day"
  >
    <div>
      Selected Day: 
    </div>
    <div
      className="Weather-houritems"
    />
  </div>
</div>
`;

exports[`renders snapshot with props 1`] = `
<div
  className="Weather-container"
>
  <div
    className="Weather-dayitems"
  >
    <div>
      No results found
    </div>
  </div>
  <div
    className="Weather-day"
  >
    <div>
      Selected Day: 
    </div>
    <div
      className="Weather-houritems"
    />
  </div>
</div>
`;
